### ISSUE TITLE
Wishlist: Resistance distance

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1742]
**What is the feature or improvement you would like to see?**

Imagine that each edge of a graph represents an electric resistor, with resistance given by the edge weight. Then the resistance distance between two vertices is the effective resistance between them.

This requires only linear algebra, preferably using sparse matrices. igraph's `sparsemat` functionality is likely sufficient for an implementation.

**Use cases for the feature**

This is useful not only for computing resistances, but for fluid flow problems as well, the mathematics of which is analogous to that of resistor networks. It will also open the door to implementing current-flow-based centrality measures: https://doi.org/10.1007/978-3-540-31856-9_44

**Notes**

Given that this is based on linear algebra, it should be carefully considered whether it's better implemented in high-level interfaces. It should be definitely prototypes in a high-level language first. However, it it can be moved to the C core without too much trouble using `sparsemat` / CSparse, that would be useful.

**References**

 - https://en.wikipedia.org/wiki/Resistance_distance
 - https://mathematica.stackexchange.com/a/195039/12 (my implementation in Mathematica)
 - https://mathematica.stackexchange.com/a/193130/12 (Henrik Schumacher's implementation in Mathematica)

**Comment by guyroznb** [comment]
I want to try to implement this. Can you assignee me?

**Comment by szhorvat** [comment]
This is best prototyped in a high-level language first, as it can be done entirely using linear algebra with sparse matrices. Then we need to check if CXSparse has everything we need to do it in C.  According to [this post](https://mathematica.stackexchange.com/a/193130/12), there are also some numerical issues to consider. This should be investigated, both in the literature and using the prototype implementation.

Are you already familiar with the background math?

**Comment by guyroznb** [comment]
Yes I am familiar with the basic math. There are some implementations in other libraries (such as [networkx](https://networkx.org/documentation/stable/_modules/networkx/algorithms/distance_measures.html#resistance_distance)) which implement resistance distance in high level language (such as python). Can we use those functions?

**Comment by ntamas** [comment]
IMHO any third-party implementation is okay as long as we can trust that it gives correct results :) The main purpose of the third-party implementation would be to make it easier to validate the results from our implementation.

**Comment by szhorvat** [comment]
We are preparing for a major change in igraph where we will move to 64-bit integers. Some fundamental function names and datatype names may be changed. This should ideally happen before the end of May.

The plan is to merge all current PRs, then go into a "freeze mode" for a while during which time the changes will be made.

It would probably be best to merge the eventual resistance distance PR only after this integer type transition is complete. I wanted to give you a warning @guyroznb, as any code written _now_ might need a bunch of tedious changes to adapt to the new integer type.

I am wary of simply copying/translating networkx's implementation. I suggest first writing up a broad plan (write down the basic math for the solution). The solution involves (possibly repeatedly) solving an underdetermined linear system: it would be good to look into the available options and their tradeoffs (accuracy and performance). Having this plan written down in plain English will also make the review much easier and quicker.
